src/execution_plan/execution_plan.c

```diff
@@ -498,304 +482,360 @@ static void _NewExecutionPlanSegment(RedisModuleCtx *ctx, GraphContext *gc,
-			uint sort_limit = (limit > 0) ? limit + skip : 0;
-			op = NewSortOp(order_expressions, direction, sort_limit);
-			Vector_Push(ops, op);
+	if(prev_projections) {
+		// We have an array of identifiers provided by a prior WITH clause -
+		// these will correspond to our first Record entities
+		uint projection_count = array_len(prev_projections);
+		for(uint i = 0; i < projection_count; i++) {
+			AR_ExpNode *projection = prev_projections[i];
+			RecordMap_FindOrAddAlias(record_map, projection->resolved_name);
 		}
+	}
 
-		if(skip_clause) {
-			OpBase *op_skip = NewSkipOp(skip);
-			Vector_Push(ops, op_skip);
-		}
+	/* Build projections from this AST's WITH, RETURN, and ORDER clauses. */
+	_ExecutionPlanSegment_BuildProjections(ast, segment, prev_projections);
 
-		if(limit_clause) {
-			OpBase *op_limit = NewLimitOp(limit);
-			Vector_Push(ops, op_limit);
-		}
-	} else if(ret_clause) {
-		// TODO we may not need a new project op if the query is something like:
-		// MATCH (a) WITH a.val AS val RETURN val
-		// Though we would still need a new projection (barring later optimizations) for:
-		// MATCH (a) WITH a.val AS val RETURN val AS e
-		if(AST_ClauseContainsAggregation(ret_clause)) {
-			op = NewAggregateOp(projections, modifies);
-		} else {
-			op = NewProjectOp(projections, modifies);
-		}
-		Vector_Push(ops, op);
+	// Extend the RecordMap to include references from clauses that do not form projections
+	// (SET, CREATE, DELETE)
+	_ExecutionPlanSegment_MapReferences(segment, ast);
 
-		if(cypher_ast_return_is_distinct(ret_clause)) {
-			op = NewDistinctOp();
-			Vector_Push(ops, op);
-		}
+	// Build query graph
+	QueryGraph *qg = BuildQueryGraph(gc, ast);
+	segment->query_graph = qg;
 
-		const cypher_astnode_t *order_clause = cypher_ast_return_get_order_by(ret_clause);
-		const cypher_astnode_t *skip_clause = cypher_ast_return_get_skip(ret_clause);
-		const cypher_astnode_t *limit_clause = cypher_ast_return_get_limit(ret_clause);
-
-		uint skip = 0;
-		uint limit = 0;
-		if(skip_clause) skip = AST_ParseIntegerNode(skip_clause);
-		if(limit_clause) limit = AST_ParseIntegerNode(limit_clause);
-
-		if(order_expressions) {
-			int direction = AST_PrepareSortOp(order_clause);
-			// The sort operation will obey a specified limit, but must account for skipped records
-			uint sort_limit = (limit > 0) ? limit + skip : 0;
-			op = NewSortOp(order_expressions, direction, sort_limit);
-			Vector_Push(ops, op);
-		}
+	// Build filter tree
+	FT_FilterNode *filter_tree = AST_BuildFilterTree(ast, record_map);
+	segment->filter_tree = filter_tree;
 
-		if(skip_clause) {
-			OpBase *op_skip = NewSkipOp(skip);
-			Vector_Push(ops, op_skip);
-		}
+	// Instantiate the Vector that will hold all constructed operations.
+	// TODO This vector could be removed with some logical changes.
+	uint clause_count = cypher_ast_query_nclauses(ast->root);
+	Vector *ops = NewVector(OpBase *, clause_count);
```

Comment:
Can we take this opportunity and get rid of Vector, use array instead.
Or even better add each operation directly to the execution-plan completely removing this intermediate container

Reply:
My original plan was to add each operation directly to the plan, but it started to feel a little flimsy when I was writing it (lots of conditional logic to handle a NULL root op, and some clumsy tracking of the head and tail of the op chain).

Switching to an array would be trivial in effort, but maybe I'll take another stab at directly updating the ExecutionPlan as well!